torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+        else:
+            # Carrying around information as member variables vs.
+            # checking one entry of the owned Tensors is annoying
+            # and error-prone. Carrying around an is_empty attribute
+            # to hide the fact that we carry around a list with a
+            # single empty Tensor is also annoying and error-prone.
+            # Both are not worth it for a minor feature.
+            raise ValueError("We do not support empty lists for now.")
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        if attr == 'requires_grad':
+            return self.tensors[0].requires_grad
+        if attr == 'grad_fn':
+            # We're assuming len(self.tensors) is non-zero
+            if self.tensors[0].grad_fn is None:
+                for i in range(1, len(self.tensors)):
+                    assert self.tensors[i].grad_fn is None
+                return None
+
+            def create_grad_fn(nested_tensor):
+                def _func(*args, **kwargs):
+                    for tensor in nested_tensor.tensors:
+                        tensor.grad_fn(*args, **kwargs)
+                return _func
+            return create_grad_fn(self)
+        if attr == 'grad':
+            grads = []
+            for tensor in self.tensors:
+                grads.append(tensor.grad)
+            if all(grad is None for grad in grads):
+                return None
+            else:
+                return NestedTensor(grads)
+        if attr == 'data':
+            data = []
+            for tensor in self.tensors:
+                data.append(tensor.data)
+            return NestedTensor(data)
+        return super().__getattribute__(attr)
+
+    def __len__(self):
+        return len(self.tensors)
+
+    def __bool__(self):
+        raise NotImplementedError(
+            "This has not been covered by NestedTensor 0.0.1")
+
+    def __str__(self):
+        tensors = self.unbind()
+        result = "nestedtensor([\n"
+        for tensor in tensors:
+            result += "  " + tensor.__str__() + "\n"
+        result += "])"
+        return result
+
+    def __repr__(self):
+        tensors = self.unbind()
+        result = "nestedtensor([\n"
+        for tensor in tensors:
+            result += "  " + tensor.__repr__() + "\n"
+        result += "])"
+        return result
+
+    def is_empty(self):
+        # This condition can never be true, since we disallow an empty list for now.
+        raise ValueError("self.tensors cannot be empty under current constraints.")
+
+    def __loop__apply(self, fn):
+        tensors = []
+        for tensor in self.tensors:
+            tensors.append(fn(tensor))
+        return tensors
+
+    def nested_size(self):
+        tensors = self.unbind()
+        sizes = []
+        for tensor in tensors:
+            sizes.append(tensor.size())
+        return tuple(sizes)
```

Comment:
Perhaps https://github.com/pytorch/pytorch/pull/22783/files#r305055454 is just frittering, but here a tuple comprehension is not only shorter, but more efficient, as it skips allocating the otherwise pointless list: `return (t.size() for t in self.unbind())`